### DISCUSSION TITLE
Album sharing on selfhosted ente

### DISCUSSION THREAD

**Discussion by technowhizz** [ente-io/ente#1421] [Q&A]
I've been playing around with the selfhosted version of ente and I'm not quite sure how the image/album sharing is supposed to work.

When sharing an album I get given an albums.ente.io link as opposed to a link to the selfhosted version.

Is album sharing available for selfhosted instances?

Thank you!

**Comment by mnvr** [comment]
> Is album sharing available for selfhosted instances?

Yes.

You'll need to start two web apps, and tell them to use your web app instead of `albums.ente.io` when creating a public link by using the `NEXT_PUBLIC_ENTE_ALBUMS_ENDPOINT` environment variable. For more details, please see this [FAQ entry](https://help.ente.io/self-hosting/faq/sharing) I recently added.

**Reply by mnvr** [reply]
Note, one thing is that the generated URLs currently have `albums.ente.io` hardcoded. We'll add a setting to change that automatically when the configuration variables are defined, but for now what you can do (after you run the above setup) is to replace `albums.ente.io` in the URL with whatever is the URL of the albums app (i.e.  the `NEXT_PUBLIC_ENTE_ALBUMS_ENDPOINT` value).

**Reply by b8ky** [reply]
Why exactly does it need to be a second web app instance, couldn't it just be one web app instance that has the public albums endpoint pointing to itself?

**Reply by mnvr** [reply]
The photos app and the public albums app are separate apps conceptually. e.g. photos requires authentication while public albums are, well, public. Right now, they happen to be share the same `app/` folder but that is something we're planning to change too, because by having them be distinct apps that are deployed separately, with separate CSP settings etc, and we can optimize the public albums app bundle to be leaner and faster to load.

**Comment by appdesign1987** [comment]
Sorry to resurect this...But i am having problems with the albums sharing part using the selfhosted version of Ente Photos.
No mather what i do if i choose to create an public link it's always coming up with a login form. I tought public link option was just created so that others can view photos without the needing to create an account ? 


I used the quickstart.sh

**Reply by mnvr** [reply]
If it redirecting to the login form then likely the albums app is missing the `NEXT_PUBLIC_ENTE_ALBUMS_ENDPOINT` environment variable.

The way it works is (currently at least) -
* The normal photos app, and the public albums app, both of them are built from the same source code
* However, when the public albums app is built, it is expected to define `NEXT_PUBLIC_ENTE_ALBUMS_ENDPOINT`, either when building/using the docker image, or in the .env file - https://github.com/ente-io/ente/blob/86007ba606de4aaeef496c813556b3c6874be387/web/apps/photos/.env#L62
* At runtime, the code looks to see if that environment variable is defined. If so, instead of serving the normal photos app, it serves the public albums app - https://github.com/ente-io/ente/blob/86007ba606de4aaeef496c813556b3c6874be387/web/apps/photos/src/pages/index.tsx#L44-L49

In the scenario you're describing, it seems that this check is not working because the `currentURL.host != albumsURL.host`. Defining `NEXT_PUBLIC_ENTE_ALBUMS_ENDPOINT` should do the trick.

You'll also need to set the same value in the museum configuration - https://ente.io/help/self-hosting/installation/config#app-endpoints. 

It might be a good idea to generally search help docs for `NEXT_PUBLIC_ENTE_ALBUMS_ENDPOINT`, it would have other info etc.

**Reply by dinobagio** [reply]
Hi,
Allow me to profit from this discussion,
I need the same help in order to make embed feature work on self hosted instance,
So far, I just did this in `museum.yaml` :
```
apps:
    # Default is https://albums.ente.io
    #
    # If you're running a self hosted instance and wish to serve public links,
    # set this to the URL where your albums web app is running.
    public-albums: https://albums.my.tld
    embed-albums: https://embed.my.tld
```
Nevertheless, every time I try to access it I'm sent back to login form, as @appdesign1987 described [but for album in its case]

**Reply by mnvr** [reply]
That might point at some other issue perhaps. Because the embed albums app doesn't need any extra NEXT_PUBLIC_ like variable, it only serves the embed album app.
https://github.com/ente-io/ente/blob/d93e5020d21c3e6ff8172b45036ed75830cd8e46/web/apps/embed/src/pages/index.tsx#L130-L142

So if it is not working, my guess is https://embed.my.tld is not actually serving the embed app, but the public albums app.

**Reply by dinobagio** [reply]
@mnvr thanks for helping.
"my guess is https://embed.my.tld/ is not actually serving the embed app, but the public albums app."
Actually I'm missing a bit the "app" notion, like I read earlier "You'll need to start two web apps".
I don't quite understand this.
I did nothing more than adding what I said in post.
How should I "add an app" ?

**Reply by mnvr** [reply]
So each of the folders in the web/apps directory is an "app"
<img width="545" height="963" alt="image" src="https://github.com/user-attachments/assets/33e901c1-95e3-43fd-8bfd-f7acebfe6521" />

These are independent web "apps" (you can think of them as "websites") that can be independently served from your web server. The default Docker image includes all of these, they are just mapped to different ports

https://github.com/ente-io/ente/blob/dc29993c5d72471bb484b68554d6a5bbd455d76a/web/Dockerfile#L31-L64

Currently the embed app is not part of the default docker image (I can't recall why, this was maybe just an oversight?, we'll try to add it for the next Docker image!), so to serve the embed app you'll need to have to build it using `yarn build:embed`, and then add it to your web server.

So by "my guess is https://embed.my.tld/ is not actually serving the embed app, but the public albums app." what I mean is that your reverse proxy (e.g. Caddy, or whatever else you're using to serve the Docker image) is mapping `https://embed.my.tld/` to port 3000 (which is photos) or port 3002 (which is the public albums app) instead of serving the "embed" app.

Unfortunately this is more complicated than it needs to be because the embed app is not part of the docker image yet. If it were, then this would've just been as simple as using port 300x (whatever the embed app will use) when serving embed.my.tld

**Reply by dinobagio** [reply]
@mnvr thanks for the explanation,
following https://ente.io/help/self-hosting/installation/env-var#ports
I've setup my haproxy to map embed to 3002 (album app).
I understand now.
And that's what I thought, EMBED is not in the docker image.
Tried to ask the discord bot about it but it would always send me back to the self hosted doc 😄 .
I'll try the YARN way.

**Reply by dinobagio** [reply]
However @mnvr I'd be very interested in knowing if and when ENTE plan to deliver EMBED in the docker image.

**Reply by mnvr** [reply]
We'll try to get it in the next image itself! A new image is released every wednesday.

**Reply by dinobagio** [reply]
@mnvr 
I've tried to yarn build and run the app (with python web server in the directory) and it worked fine.
I've been able to test the feature.
Thank you so much.
Looking forward to getting the new image, it will make things easier eventually.

**Reply by mnvr** [reply]
Awesome!

**Reply by dinobagio** [reply]
> We'll try to get it in the next image itself! A new image is released every wednesday.

All good. Thank you so much.

**Reply by Magneticdud** [reply]
hi, when this is implemented, do i need to add an environment variable to the docker image for web, or it just takes it from museum.yaml?

edit: i experimented, and added apps.embed-albums to the museum.yaml (as the documentation says to do) and the web app correctly generated the embed code with my url. Problem is that it says error 404 when i open the iframe, but it might be that it's simply missing from the docker image yet.

**Comment by appdesign1987** [comment]
Hi and thank you all for the reply's and sorry for my late reactiopn..Bussy family live over here in Holland :-).

This is my docker-compose file:
  _web:
    image: ghcr.io/ente-io/web
    # Uncomment what you need to tweak.
    ports:
      - 3000:3000 # Photos web app
      - 3001:3001 # Accounts
      - 3002:3002 # Public albums
      - 3003:3003 # Auth
      - 3004:3004 # Cast
    # Modify these values to your custom subdomains, if using any
    environment:
      ENTE_API_ORIGIN: http://api.ente.jeroenvd.nl
      ENTE_ALBUMS_ORIGIN: https://albums.ente.jeroenvd.nl
      EXT_PUBLIC_ENTE_ALBUMS_ENDPOINT: https://albums.ente.jeroenvd.nl_

``

and this is the app section of my museum.yaml file:

_apps:
    # If you're running a self hosted instance and wish to serve public links,
    # set this to the URL where your albums web app is running.
    public-albums: https://albums.ente.jeroenvd.nl
    cast: https://cast.ente.jeroenvd.nl
    embed-albums: https://embed.ente.jeroenvd.nl
    # Set this to the URL where your accounts web app is running, primarily used for
    # passkey based 2FA.
    accounts: https://auth.ente.jeroenvd.nl_


So as you can see i have specified the public-albums url (public.ente.jeroenvd.nl) in museum.yaml and in my docker compose file.
But still when i go to create an public link in Ente Photo the url still begins with web.ente instead of albums.ente.jeroenvd.nl
So there something goint wrong with the docker web container itself. Unfortunately i have not had enough time to look further in the container to see what's wron.

**Reply by appdesign1987** [reply]
> Hi and thank you all for the reply's and sorry for my late reactiopn..Bussy family live over here in Holland :-).
> 
> This is my docker-compose file: _web: image: ghcr.io/ente-io/web # Uncomment what you need to tweak. ports: - 3000:3000 # Photos web app - 3001:3001 # Accounts - 3002:3002 # Public albums - 3003:3003 # Auth - 3004:3004 # Cast # Modify these values to your custom subdomains, if using any environment: ENTE_API_ORIGIN: http://api.ente.jeroenvd.nl ENTE_ALBUMS_ORIGIN: https://albums.ente.jeroenvd.nl EXT_PUBLIC_ENTE_ALBUMS_ENDPOINT: https://albums.ente.jeroenvd.nl_
> 
> ``
> 
> and this is the app section of my museum.yaml file:
> 
> _apps: # If you're running a self hosted instance and wish to serve public links, # set this to the URL where your albums web app is running. public-albums: https://albums.ente.jeroenvd.nl cast: https://cast.ente.jeroenvd.nl embed-albums: https://embed.ente.jeroenvd.nl # Set this to the URL where your accounts web app is running, primarily used for # passkey based 2FA. accounts: https://auth.ente.jeroenvd.nl_
> 
> So as you can see i have specified the public-albums url (public.ente.jeroenvd.nl) in museum.yaml and in my docker compose file. But still when i go to create an public link in Ente Photo the url still begins with web.ente instead of albums.ente.jeroenvd.nl So there something goint wrong with the docker web container itself. Unfortunately i have not had enough time to look further in the container to see what's wron.


After some searching i have it working. Altough i need to change the public url i got when creating a public link in Ente Photos. If i change that URL to albums.ente it's working without any problems..
